### PR TITLE
Add compilable foreach RAdam support

### LINKED ISSUE #117807: Supporting full graph compilation of RAdam
### 🚀 The feature, motivation and pitch

Currently RAdam graph breaks on data-dependent control flow https://github.com/pytorch/pytorch/blob/a468b9fbdf75b2da128e5fac4284ebbb981cd31a/torch/optim/radam.py#L326

We should be able to handle this graph break while compiling with `torch.where` and avoid the graph break. 
After implementing this add a test at https://github.com/pytorch/pytorch/blob/a468b9fbdf75b2da128e5fac4284ebbb981cd31a/test/inductor/test_compiled_optimizers.py

### Alternatives

_No response_

### Additional context

_No response_

cc @vincentqb @jbschlosser @albanD @janeyx99 @crcrpar @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### PR DESCRIPTION
**Pull request by mlazos**
Fixes https://github.com/pytorch/pytorch/issues/117807

This brings the number of supported optimizers with `torch.compile` to 11/13 (!)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: torch/testing/_internal/common_optimizers.py [hunk 1]
```diff
@@ -599,6 +599,9 @@ def optim_error_inputs_func_nadam(device, dtype):
 
 # Weird story bro, NAdam and RAdam do not have maximize.
 def optim_inputs_func_radam(device=None):
+    cuda_supported_configs = [
+        OptimizerInput(params=None, kwargs={"capturable": True}, desc="capturable"),
+    ]
```
Comment (janeyx99) [inline torch/testing/_internal/common_optimizers.py:604]: It'd be good to add one for weight decay + another for weight decay + decoupled weight decay

Comment (mlazos) [reply torch/testing/_internal/common_optimizers.py:604]: Oh does it matter that these are covered in the cudagraph tests?

Comment (mlazos) [reply torch/testing/_internal/common_optimizers.py:604]: Discussed offline, I can add

File: torch/_dynamo/eval_frame.py [hunk 2]
```diff
@@ -1525,9 +1526,9 @@ def patch():
         # Note: we don't support sparsity, data-dependent control, or tracing through backwards
```
Comment (janeyx99) [inline torch/_dynamo/eval_frame.py:1526]:
```suggestion
        # Note: we don't support sparsity or tracing through backwards
```

File: torch/optim/radam.py [hunk 3]
```diff
@@ -283,6 +292,7 @@ def _single_tensor_radam(
     eps: float,
     differentiable: bool,
     decoupled_weight_decay: bool,
+    capturable: bool,
```
Comment (janeyx99) [inline torch/optim/radam.py:295]: Should error if this is true, since this isn't supported, similar to ASGD, right?

File: torch/testing/_internal/common_optimizers.py [hunk 4]
```diff
@@ -599,6 +599,26 @@ def optim_error_inputs_func_nadam(device, dtype):
...
+            desc="capturable",
+        ),
+        OptimizerInput(
+            params=None,
+            kwargs={
+                "capturable": True,
+                "weight_decay": 0.9,
+                "decoupled_weight_decay": True,
+            },
+            desc="capturable",
```
Comment (janeyx99) [inline torch/testing/_internal/common_optimizers.py:619]:
```suggestion
            desc="capturable, weight_decay, decoupled_weight_decay",
```

File: torch/testing/_internal/common_optimizers.py [hunk 5]
```diff
@@ -599,6 +599,26 @@ def optim_error_inputs_func_nadam(device, dtype):
...
 def optim_inputs_func_radam(device=None):
+    cuda_supported_configs = [
+        OptimizerInput(params=None, kwargs={"capturable": True}, desc="capturable"),
+        OptimizerInput(
+            params=None,
+            kwargs={
+                "capturable": True,
+                "weight_decay": 0.9,
+            },
+            desc="capturable",
```
Comment (janeyx99) [inline torch/testing/_internal/common_optimizers.py:610]: update this as well

File: torch/testing/_internal/common_optimizers.py [hunk 6]
```diff
@@ -599,6 +599,26 @@ def optim_error_inputs_func_nadam(device, dtype):
 
 # Weird story bro, NAdam and RAdam do not have maximize.
 def optim_inputs_func_radam(device=None):
+    cuda_supported_configs = [
+        OptimizerInput(params=None, kwargs={"capturable": True}, desc="capturable"),
+        OptimizerInput(
+            params=None,
+            kwargs={
+                "capturable": True,
+                "weight_decay": 0.9,
```
Comment (janeyx99) [inline torch/testing/_internal/common_optimizers.py:608]:
```suggestion
                "weight_decay": 0.1,
```
This is much more common

File: torch/testing/_internal/common_optimizers.py [hunk 7]
```diff
@@ -599,6 +599,26 @@ def optim_error_inputs_func_nadam(device, dtype):
...
+            },
+            desc="capturable",
+        ),
+        OptimizerInput(
+            params=None,
+            kwargs={
+                "capturable": True,
+                "weight_decay": 0.9,
```
Comment (janeyx99) [inline torch/testing/_internal/common_optimizers.py:616]: same here

File: torch/optim/radam.py (hunk 3 above)
Comment (janeyx99) [reply torch/optim/radam.py:295]: Wait...the foreach_matches_forloop eager tests pass though, so is this supported on single tensor already in eager?

Comment (mlazos) [reply torch/optim/radam.py:295]: Since we don't have an error it passes.

File: torch/optim/radam.py [hunk 8]
```diff
@@ -405,29 +433,61 @@ def _multi_tensor_radam(
...
+            torch._foreach_mul_(num, sub2)
+            torch._foreach_mul_(num, rho_inf)
+            rho_inf = ((rho_inf - 4) * (rho_inf - 2))
+            denom = torch._foreach_mul(rho_t_list, rho_inf)
+            torch._foreach_div_(num, denom)
+            torch._foreach_sqrt_(num)
+
+            # TODO(mlazos): we should try and get a foreach_where op https://github.com/pytorch/pytorch/issues/117884
+            rect = [torch.where(rho_t > 5, n, torch.zeros_like(n)) for n, rho_t in zip(num, rho_t_list)]
+            unrect_step_size = [torch.where(rect > 0, torch.zeros_like(rect), torch.ones_like(rect)) for rect in rect]
```
Comment (vadimkantorov) [inline torch/optim/radam.py:448]: btw, torch.where now supports python scalars, so probably `[torch.where(rect > 0, 0.0, 1.0) for rect in rect]` should work or even `(rect > 0).type_as(rect)`

Comment (mlazos) [reply torch/optim/radam.py:448]: I actually need a full tensor of the value for the usage below actually

Comment (vadimkantorov) [reply torch/optim/radam.py]: The output of torch.where is still a full tensor, even if you pass in scalars instead of zeros_like/ones_like, right?

Currently zeros_like and ones_like are materialized just for passing into torch.where

File: torch/optim/radam.py [hunk 9]
```diff
@@ -8,11 +8,11 @@
     Optimizer,
     _default_to_fused_or_foreach,
     _differentiable_doc,
+    _capturable_doc,
     _dispatch_sqrt,
```
Comment (vadimkantorov) [inline torch/optim/radam.py:12]: Also, could `_dispatch_sqrt` be replaced by `** 0.5`? At least functionally, this should be identical, not sure about the speed...

File: torch/optim/radam.py (hunk 8 above)
Comment (mlazos) [reply torch/optim/radam.py]: Ah you're right, done

File: torch/optim/radam.py (hunk 9 above)
Comment (mlazos) [reply torch/optim/radam.py:12]: I think you're right, but I'll do that update in a separate PR and update all optimizers

Comment (vadimkantorov) [reply torch/optim/radam.py:12]: It would be nice to be sure if `** 0.5` maps exactly to `torch.sqrt` implementation and not some generic `torch.pow`...

File: torch/optim/radam.py (hunk 8 above)
Comment (vadimkantorov) [reply torch/optim/radam.py]: and maybe even `(rect <= 0).type_as(rect)` would be correct here... I don't know if any of these is better for torch.compile...

File: test/test_optim.py [hunk 10]
```diff
@@ -394,6 +410,11 @@ def test_peak_memory_foreach(self, device, dtype, optim_info):
                     # bias_correction, mus, and mu_nexts
                     nintermediates = 5
 
+                if optim_cls.__name__ == "RAdam":
+                    # RAdam has six intermediates with capturable
```
Comment (janeyx99) [inline test/test_optim.py:414]: is this the min we could go

File: torch/optim/radam.py [hunk 11]
```diff
@@ -60,13 +62,11 @@ def __setstate__(self, state):
...
-            state_values[0]["step"]
-        )
-        if not step_is_tensor:
-            for s in state_values:
-                s["step"] = torch.tensor(float(s["step"]), dtype=_get_scalar_dtype())
+            group.setdefault("capturable", False)
+            for p in group["params"]:
+                p_state = self.state.get(p, [])
+                if len(p_state) != 0 and not torch.is_tensor(p_state['step']):
+                    p_state["step"] = torch.tensor(float(p_state["step"]), dtype=_get_scalar_dtype())
```
Comment (janeyx99) [inline torch/optim/radam.py:69]: step needs to be on the param's device if capturable is on and on CPU otherwise

File: test/test_optim.py (hunk 10 above)
Comment (mlazos) [reply test/test_optim.py]: Lowered to 4 after @vadimkantorov 's suggestions!

File: torch/optim/radam.py (hunk 11 above)
Comment (mlazos) [reply torch/optim/radam.py:69]: I copied this from Adam?

Comment (mlazos) [reply torch/optim/radam.py:69]: this is the non-tensor case. So I don't think this applies right? (at least I'm guessing)

Comment (janeyx99) [reply torch/optim/radam.py:69]: no this applies, it should get promoted to a tensor correctly. curr adam is bad and is getting fixed in my pr

File: torch/optim/radam.py [hunk 12]
```diff
@@ -379,16 +396,29 @@ def _multi_tensor_radam(
         if grouped_state_steps[0].is_cpu:
             torch._foreach_add_(grouped_state_steps, torch.tensor(1.0, device='cpu'), alpha=1.0)
         else:
-            torch._foreach_add_(grouped_state_steps, 1)
+            torch._foreach_add_(grouped_state_steps, 1.0)
```
Comment (janeyx99) [inline torch/optim/radam.py:399]: wait why is this necessary

File: torch/testing/_internal/common_optimizers.py [hunk 13]
```diff
@@ -1487,6 +1514,48 @@ def _get_optim_inputs_including_global_cliquey_kwargs(
...
+            DecorateInfo(
+                skipIfTorchDynamo(
+                    "See https://github.com/pytorch/pytorch/issues/116494"
+                ),
+                "TestOptimRenewed",
+                "test_state_dict_deterministic",
+            ),
+            DecorateInfo(
+                skipIfTorchDynamo(
+                    "Should be fixed by https://github.com/pytorch/pytorch/issues/118230"
```
Comment (janeyx99) [inline torch/testing/_internal/common_optimizers.py:1543]: Adding a capturable impl won't fix this because capturable doesn't work for cpu. Did you mean to link another issue?

File: torch/testing/_internal/common_optimizers.py [hunk 14]
```diff
@@ -1459,6 +1479,13 @@ def _get_optim_inputs_including_global_cliquey_kwargs(
                 "TestOptimRenewed",
                 "test_forloop_goes_right_direction",
             ),
+            DecorateInfo(
+                unittest.skip(
+                    "RAdam does not support capturable single tensor https://github.com/pytorch/pytorch/issues/118230"
```
Comment (janeyx99) [inline torch/testing/_internal/common_optimizers.py:1484]: I think this should be a skip in the test

File: torch/optim/radam.py (hunk 12 above)
Comment (mlazos) [reply torch/optim/radam.py:399]: not necessary, I just did it since we do it in most other places

Comment (mlazos) [reply torch/optim/radam.py:399]: if you care, I can remove it.

File: torch/optim/radam.py (hunk 11 above)
Comment (mlazos) [reply torch/optim/radam.py:69]: I'll update

File: torch/testing/_internal/common_optimizers.py (hunk 13 above)
Comment (mlazos) [reply torch/testing/_internal/common_optimizers.py:1543]: oh yeah I think I meant the init_group issue

File: torch/testing/_internal/common_optimizers.py [hunk 15]
```diff
@@ -1487,6 +1514,48 @@ def _get_optim_inputs_including_global_cliquey_kwargs(
...
+                    "See https://github.com/pytorch/pytorch/issues/115607"
+                ),
+                "TestOptimRenewed",
+                "test_foreach_matches_forloop",
+            ),
+            DecorateInfo(
+                toleranceOverride(
+                    {
+                        # previously atol=1e-7, rtol=1e-7
+                        torch.float64: tol(atol=1.5e-7, rtol=1.1e-7)
```
Comment (janeyx99) [inline torch/testing/_internal/common_optimizers.py:1528]: I don't like this...do you have an explanation why?

File: torch/testing/_internal/common_optimizers.py (hunk 14 above)
Comment (mlazos) [reply torch/testing/_internal/common_optimizers.py:1484]: So in this case, I made it an actual skip. Since we are testing forloop directly, I felt like it made sense to fully skip here vs redundant tests with capturable=True and capturable=False

File: torch/testing/_internal/common_optimizers.py (hunk 15 above)
Comment (mlazos) [reply torch/testing/_internal/common_optimizers.py:1528]: not really, other than there are a lot of ops in RAdam capturable so I think there are most casts vs for loop. These are really close I don't think there's an issue.

File: torch/testing/_internal/common_optimizers.py (hunk 14 above)
Comment (janeyx99) [reply torch/testing/_internal/common_optimizers.py:1484]: With this skip, the noncapturable configs don't get run though, and that's worse for coverage.

File: torch/optim/radam.py [hunk 16]
```diff
@@ -405,29 +435,67 @@ def _multi_tensor_radam(
...
+            torch._foreach_div_(unrect_step_size, bias_correction1)
+            torch._foreach_neg_(unrect_step_size)
+
+            bias_correction2 = torch._foreach_pow(beta2, grouped_state_steps)
+            torch._foreach_neg_(bias_correction2)
+            torch._foreach_add_(bias_correction2, 1)
+            torch._foreach_sqrt_(bias_correction2)
+            torch._foreach_mul_(bias_correction2, lr)
+            torch._foreach_mul_(bias_correction2, rect)
+            del rect
```
[the remark this answers is not in the document]
Comment (mlazos) [reply torch/optim/radam.py:472]: I'll leave it unless you want me to remove, I think an early `del` is still a tiny bit better just for mem in general.

File: torch/optim/radam.py (hunk 12 above)
Comment (janeyx99) [reply torch/optim/radam.py:401]: Where else do you see 1.0?

I'd prefer this not be here if there's no reason

File: torch/optim/radam.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (janeyx99) [reply torch/optim/radam.py:472]: oh hm i think this actually is necessary cuz you introduce buffer later

we can leave it

File: torch/optim/radam.py (hunk 12 above)
Comment (mlazos) [reply torch/optim/radam.py:401]: There is an instance with ASGD as well, but I'll just remove idc

**Review by janeyx99 — APPROVED** [review]
This approve is conditional on green CI (as always) and I suspect you will need more in-test skips before that happens.

I am reluctant to approve and am doing so following our agreement earlier this week where I said skips were acceptable in the interim. Since then, I've come to regret ever allowing a divergence of capturable support between the forloop/single-tensor + foreach/multi-tensor implementations (for ASGD and Adamax) and am not happy with this intermediate state we would leave ourselves in. 
- The lack of support for our simplest implementation is bad for consistency and expectations from users
- We will likely want compiled forloop for compiled optimizer in backward
- It is not much more work to do it now (~30%+) than doing it from scratch when we will need it later

So I would _still_ prefer having this PR include both implementations supporting `capturable` even now to avoid the time we'd be in the intermediate state, and there is no rush from my end to get this in right now vs when you'd have the time to work on single tensor in a week or so.

File: torch/testing/_internal/common_optimizers.py (hunk 15 above)
Comment (janeyx99) [reply torch/testing/_internal/common_optimizers.py:1521]: We should check if this is still applicable once the single tensor capturable happens.
